lib/hirefire/macro/sidekiq.rb

```diff
@@ -28,35 +28,77 @@ def queue(*queues)
           options = {}
         end
 
-        queues = queues.map(&:to_s)
-        queues = ::Sidekiq::Stats.new.queues.map { |name, _| name } if queues.empty?
+        queues.map!(&:to_s)
+        all_queues = ::Sidekiq::Queue.all.map(&:name)
+        queues = all_queues if queues.empty?
 
+        if fast_lookup_capable?(queues, all_queues)
+          fast_lookup(options)
+        else
+          dynamic_lookup(queues, options)
+        end
+      end
+
+      private
+
+      def fast_lookup_capable?(queues, all_queues)
+        # When no queue names are provided (or all of them are), we know we
+        # can peform much faster counts using Sidekiq::Stats and Redis
+        queues.sort == all_queues.sort
+      end
+
+      def fast_lookup(options)
+        stats = ::Sidekiq::Stats.new
+
+        in_queues = stats.enqueued
+
+        if !options[:skip_scheduled]
+          in_schedule = ::Sidekiq.redis { |c| c.zcount('schedule', '-inf', Time.now.to_f) }
+        end
+
+        if !options[:skip_retries]
+          in_retry = ::Sidekiq.redis { |c| c.zcount('retry', '-inf', Time.now.to_f) }
+        end
+
+        in_progress = stats.workers_size
+
+        [in_queues, in_schedule, in_retry, in_progress].compact.inject(&:+)
+      end
+
+      def dynamic_lookup(queues, options)
         in_queues = queues.inject(0) do |memo, name|
           memo += ::Sidekiq::Queue.new(name).size
           memo
         end
 
         if !options[:skip_scheduled]
           max = options[:max_scheduled]
+
+          # For potentially long-running loops, compare all jobs against
+          # time when the set snapshot was taken to avoid incorrect counts.
+          now = Time.now
+
           in_schedule = ::Sidekiq::ScheduledSet.new.inject(0) do |memo, job|
-            memo += 1 if queues.include?(job["queue"]) && job.at <= Time.now
+            memo += 1 if queues.include?(job["queue"]) && job.at <= now
             break memo if max && memo >= max
             memo
           end
         end
 
         if !options[:skip_retries]
+          now = Time.now
+
           in_retry = ::Sidekiq::RetrySet.new.inject(0) do |memo, job|
-            memo += 1 if queues.include?(job["queue"]) && job.at <= Time.now
+            memo += 1 if queues.include?(job["queue"]) && job.at <= now
             memo
           end
         end
 
-        i = ::Sidekiq::VERSION >= "3.0.0" ? 2 : 1
-        in_progress = ::Sidekiq::Workers.new.inject(0) do |memo, job|
-          memo += 1 if queues.include?(job[i]["queue"]) && job[i]["run_at"] <= Time.now.to_i
-          memo
-        end
+        # Objects yielded to Workers#each:
+        # https://github.com/mperham/sidekiq/blob/305ab8eedc362325da2e218b2a0e20e510668a42/lib/sidekiq/api.rb#L912
+        in_progress = ::Sidekiq::Workers.new.select do |key, tid, job|
+          queues.include?(job['queue']) && job['run_at'] <= Time.now.to_i
```

Comment:
Perhaps extract `Time.now.to_i` into a separate variable.

```rb
now = Time.now.to_i
```

And use that here instead like you did with the other examples.

Reply:
@mrrooijen my thinking was that scheduled and retries are likely to be long-running. If we only assign the `now` variable once at the beginning, a long time can pass while we count scheduled, so by the time we get to retries, they're pretty much guaranteed to return 0 because we're comparing them to a pretty old timestamp. With that in mind, we can define it for the workers loop as well, I just didn't think it's common to have a significant number of workers at the same time.

Reply:
Right. It's more that it performs multiple `Time.now.to_i` calls rather than just once, which would probably be sufficient. If you have 1000 workers operating at the same time, then that's 1000x `Time.now.to_i` per macro call, while a single invocation would probably be sufficient.

Reply:
Cool @mrrooijen pushed an update, thanks for that 👍